mrmustard/lab/gates.py

```diff
@@ -430,93 +430,92 @@ def _validate_modes(self, modes):
 class Interferometer(Parametrized, Transformation):
     r"""N-mode interferometer.
 
-    It corresponds to a Ggate with zero mean and a ``2N x 2N`` orthogonal symplectic matrix.
+    It corresponds to a Ggate with zero mean and a ``2N x 2N`` unitary symplectic matrix.
 
     Args:
         num_modes (int): the num_modes-mode interferometer
-        orthogonal (2d array): a valid orthogonal matrix. For N modes it must have shape `(2N,2N)`
-        orthogonal_trainable (bool): whether orthogonal is a trainable variable
+        unitary (2d array): a valid unitary matrix. For N modes it must have shape `(2N,2N)`
+        unitary_trainable (bool): whether unitary is a trainable variable
         modes (optional, List[int]): the list of modes this gate is applied to
     """
 
     def __init__(
         self,
         num_modes: int,
-        orthogonal: Optional[Tensor] = None,
-        orthogonal_trainable: bool = False,
+        unitary: Optional[Tensor] = None,
+        unitary_trainable: bool = False,
         modes: Optional[List[int]] = None,
     ):
         if modes is not None and (
             num_modes != len(modes) or any(mode >= num_modes for mode in modes)
         ):
             raise ValueError("Invalid number of modes and the mode list here!")
-        if orthogonal is None:
-            U = math.random_unitary(num_modes)
-            orthogonal = math.block([[math.real(U), -math.imag(U)], [math.imag(U), math.real(U)]])
+        if unitary is None:
+            unitary = math.random_unitary(2 * num_modes)
         super().__init__(
-            orthogonal=orthogonal,
-            orthogonal_trainable=orthogonal_trainable,
+            unitary=unitary,
+            unitary_trainable=unitary_trainable,
         )
         self._modes = modes or list(range(num_modes))
         self.is_gaussian = True
 
     @property
     def X_matrix(self):
-        return self.orthogonal.value
+        return self.unitary.value
 
     def _validate_modes(self, modes):
-        if len(modes) != self.orthogonal.value.shape[-1] // 2:
+        if len(modes) != self.unitary.value.shape[-1] // 2:
             raise ValueError(
-                f"Invalid number of modes: {len(modes)} (should be {self.orthogonal.shape[-1] // 2})"
+                f"Invalid number of modes: {len(modes)} (should be {self.unitary.shape[-1] // 2})"
             )
 
     def __repr__(self):
         modes = self.modes
-        orthogonal = repr(math.asnumpy(self.orthogonal.value)).replace("\n", "")
-        return f"Interferometer(num_modes = {len(modes)}, orthogonal = {orthogonal}){modes}"
+        unitary = repr(math.asnumpy(self.unitary.value)).replace("\n", "")
+        return f"Interferometer(num_modes = {len(modes)}, unitary = {unitary}){modes}"
 
 
 class RealInterferometer(Parametrized, Transformation):
```

Comment:
I'm confused about this real interferometer. Its unitary matrix should be real? We use the same unitary update for it? But the gradient is complex for this.

Reply:
The only issue now is related to this one. Then we are done!!

Reply:
@ziofil I don't know how to deal with this RealInterferometer...

Reply:
Don't worry, we will solve it :)

Reply:
So, the RealInterferometer does not mix p and q, which means that in its symplectic matrix S = [[A,B],[C,D]] the off-diagonal blocks B and C are zero. Given that they correspond to the imaginary part of the unitary matrix that parametrizes S, it means that the unitary matrix is real (hence the name, RealInterferometer).
It could be sufficient that in the formulas that describe the parameter update you make sure no updates have an imaginary component? (I'm not 100% sure if that's sufficient, surely it's necessary though)

Reply:
In the definition of this RealInterferometer, there is only a N by N real unitary matrix (which is only unitary in the MM now). And the trainable parameter is this N by N matrix. I was wondering if I change the code to train the 2N by 2N matrix ([[M,0],[0,M]]), it would be the same as the Interferometer and there is no need to have this RealInterferometer one.

Reply:
if unitary is None:
            unitary = math.random_unitary(num_modes)
        super().__init__(unitary=unitary, unitary_trainable=unitary_trainable)

1. unitary = math.random_unitary(num_modes) -> not real, just a random unitary.
2. super().__init__(unitary=unitary, unitary_trainable=unitary_trainable), the training it on this N by N unitary, not the 2N by 2N one.

Reply:
if unitary is None:
        U = math.real(math.random_unitary(num_modes))
        unitary = math.block([[U, math.zeros_like(U)], [math.zeros_like(U), U]])

This is what I want to suggest...